src/collocated_estim.jl

```diff
@@ -0,0 +1,165 @@
+        #     nnsol = [f([û[i], u1[i]],
+        #         Tar.prob.p,
+        #         t[i])
+        #              for i in 1:length(û[:, 1])]
+        # end
+        # form of NN output matrix output dim x n
+        # nnsol = reduce(hcat, nnsol)
+
+        # > Instead of dataset gradients trying NN derivatives with dataset collocation 
+        # # convert to matrix as nnsol  
+
+        physlogprob = 0
+        for i in 1:length(Tar.prob.u0)
+            # can add phystd[i] for u[i] 
+            physlogprob += logpdf(MvNormal(deri_physsol[i, :],
+                    LinearAlgebra.Diagonal(map(abs2,
+                        (Tar.l2std[i] * 4.0) .*
+                        ones(length(nnsol[i, :]))))),
+                nnsol[i, :])
+        end
+        return physlogprob
+    else
+        return 0
+    end
+end
+
+# PDE(DU,U,P,T)=0
+
+# Derivated via Central Diff
+# function calculate_derivatives2(dataset)
+#     x̂, time = dataset
+#     num_points = length(x̂)
+#     # Initialize an array to store the derivative values.
+#     derivatives = similar(x̂)
+
+#     for i in 2:(num_points - 1)
+#         # Calculate the first-order derivative using central differences.
+#         Δt_forward = time[i + 1] - time[i]
+#         Δt_backward = time[i] - time[i - 1]
+
+#         derivative = (x̂[i + 1] - x̂[i - 1]) / (Δt_forward + Δt_backward)
+
+#         derivatives[i] = derivative
+#     end
+
+#     # Derivatives at the endpoints can be calculated using forward or backward differences.
+#     derivatives[1] = (x̂[2] - x̂[1]) / (time[2] - time[1])
+#     derivatives[end] = (x̂[end] - x̂[end - 1]) / (time[end] - time[end - 1])
+#     return derivatives
+# end
+
+function calderivatives(prob, dataset)
+    chainflux = Flux.Chain(Flux.Dense(1, 8, tanh), Flux.Dense(8, 8, tanh),
+        Flux.Dense(8, 2)) |> Flux.f64
+    # chainflux = Flux.Chain(Flux.Dense(1, 7, tanh), Flux.Dense(7, 1)) |> Flux.f64
+    function loss(x, y)
+        # sum(Flux.mse.(prob.u0[1] .+ (prob.tspan[2] .- x)' .* chainflux(x)[1, :], y[1]) +
+        #     Flux.mse.(prob.u0[2] .+ (prob.tspan[2] .- x)' .* chainflux(x)[2, :], y[2]))
+        # sum(Flux.mse.(prob.u0[1] .+ (prob.tspan[2] .- x)' .* chainflux(x)[1, :], y[1]))
+        sum(Flux.mse.(chainflux(x), y))
+    end
+    optimizer = Flux.Optimise.ADAM(0.01)
+    epochs = 3000
+    for epoch in 1:epochs
+        Flux.train!(loss,
+            Flux.params(chainflux),
+            [(dataset[end]', dataset[1:(end - 1)])],
+            optimizer)
+    end
+
+    # A1 = (prob.u0' .+
+    #   (prob.tspan[2] .- (dataset[end]' .+ sqrt(eps(eltype(Float64)))))' .*
+    #   chainflux(dataset[end]' .+ sqrt(eps(eltype(Float64))))')
+
+    # A2 = (prob.u0' .+
+    #       (prob.tspan[2] .- (dataset[end]'))' .*
+    #       chainflux(dataset[end]')')
+
+    A1 = chainflux(dataset[end]' .+ sqrt(eps(eltype(dataset[end][1]))))
+    A2 = chainflux(dataset[end]')
+
+    gradients = (A2 .- A1) ./ sqrt(eps(eltype(dataset[end][1])))
```

Comment:
You shouldn't use this for gradients?

Reply:
no, we use the gradient of the NN in new loss, ill remove this.

Reply:
not anymore, i wrote this for it to work for a specific case, normally the same is done using the derivative function NNodederi in the BPINN ode solver.